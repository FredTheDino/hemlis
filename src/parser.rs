#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    macro_rules! gen_parser {
        ($a:ident, $p:ident) => {
            pub fn $a<'t>(src: &'t str) -> String {
                pub fn inner<'t>(src: &'t str) -> Result<String, String> {
                    use crate::{ast::Ast, grammer, lexer};
                    use std::io::BufWriter;

                    let parser = grammer::$p::new();
                    let out = parser
                        .parse(
                            lexer::lex(&src)
                                .into_iter()
                                .map(|(token, span)| Ok((span.start, token?, span.end))),
                        )
                        .map_err(|x| format!("{:?}", x))?;

                    let mut buf = BufWriter::new(Vec::new());
                    out.show(0, &mut buf).map_err(|x| format!("{:?}", x))?;
                    let inner = buf.into_inner().map_err(|x| format!("{:?}", x))?;
                    String::from_utf8(inner).map_err(|x| format!("{:?}", x))
                }
                match inner(src) {
                    Ok(s) | Err(s) => s,
                }
            }
        };
    }

    gen_parser!(p_header, HeaderParser);
    gen_parser!(p_import, ImportParser);

    #[test]
    fn empty_string() {
        assert_snapshot!(p_header(""));
    }

    #[test]
    fn normal_definition() {
        assert_snapshot!(p_header("module A (a, b, c) where"));
    }

    #[test]
    fn everything_header() {
        assert_snapshot!(p_header(r#"
module A.B.C (a, class B, C, D(..), E(F, G), 
 (+), type (+), module H) where

import A as A
import A.B.C as A.C
import A.B.C (a, class B, 
 C, D(..), E(F, G), (+), type (+))
import A.B.C hiding (foo)
        "#));
    }

    #[test]
    fn simple_import() {
        assert_snapshot!(p_import("import A as A"))
    }

    #[test]
    fn import_with_newlines() {
        assert_snapshot!(p_import("import A (foo\n , bar\n , baz)"))
    }
}
