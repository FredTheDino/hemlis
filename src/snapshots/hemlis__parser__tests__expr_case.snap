---
source: src/parser.rs
expression: "p_expr(r\"\n        case 1 + 1 of\n            2 -> foo\n            3 | Just _ <- foo bar -> baz\n                    where\n                        baz = 3\n        \")"
---
33 of 33
===
Module
 Header
  MName
   (0, 7)->(0, 11)
  NULL
  Decl::Def
   Name
    (2, 0)->(2, 1)
   GuardedExpr::Unconditional
    Expr::Case
     (3, 8)->(3, 12)
      Expr::Op
       Expr::Number
        Number
         (3, 13)->(3, 14)
       QOp
        NULL
        Op
         (3, 15)->(3, 16)
       Expr::Number
        Number
         (3, 17)->(3, 18)
      CaseBranch
        Binder::Number
         Number
          (4, 12)->(4, 13)
       GuardedExpr::Unconditional
        Expr::Ident
         QName
          NULL
          Name
           (4, 17)->(4, 20)
      CaseBranch
        Binder::Number
         Number
          (5, 12)->(5, 13)
       GuardedExpr::Guarded
         (
           Guard::Binder
            Binder::App
             Binder::Constructor
              QProperName
               NULL
               ProperName
                (5, 16)->(5, 20)
             Binder::Wildcard
              (5, 21)->(5, 22)
            Expr::App
             Expr::Ident
              QName
               NULL
               Name
                (5, 26)->(5, 29)
             Expr::Ident
              QName
               NULL
               Name
                (5, 30)->(5, 33)
          Expr::Where
           (6, 20)->(6, 25)
           Expr::Ident
            QName
             NULL
             Name
              (5, 37)->(5, 40)
            LetBinding::Name
             Name
              (7, 24)->(7, 27)
             GuardedExpr::Unconditional
              Expr::Number
               Number
                (7, 30)->(7, 31)
         )
===
